lib/src/authentication/client.dart

```diff
@@ -34,17 +40,68 @@ class _AuthenticatedClient extends http.BaseClient {
     // to given serverBaseUrl. Otherwise credential leaks might ocurr when
     // archive_url hosted on 3rd party server that should not receive
     // credentials of the first party.
-    if (credential.canAuthenticate(request.url.toString())) {
+    if (credential != null &&
+        credential.canAuthenticate(request.url.toString())) {
       request.headers[HttpHeaders.authorizationHeader] =
           await credential.getAuthorizationHeaderValue();
     }
-    return _inner.send(request);
+
+    try {
+      final response = await _inner.send(request);
+      if (response.statusCode == 401) _handleError(response);
+      return response;
+    } on PubHttpException catch (error) {
+      if (error.response?.statusCode == 403) _handleError(error.response);
+      rethrow;
+    }
+  }
+
+  void _handleError(http.BaseResponse response) {
+    String serverMessage;
+    try {
+      final wwwAuthenticateHeaderValue =
+          response.headers[HttpHeaders.wwwAuthenticateHeader];
+      if (wwwAuthenticateHeaderValue != null) {
+        final parsedValue = HeaderValue.parse(wwwAuthenticateHeaderValue,
+            parameterSeparator: ',');
+        if (parsedValue.parameters['realm'] == 'pub') {
+          serverMessage = parsedValue.parameters['message'];
+        }
+      }
+    } catch (_) {
```

Comment:
Hopefully, the parsing can only throw `FormatException`, we should check this, but if so it would be preferable to do:
```suggestion
    } on FormatException {
```

Reply:
Current implementation throws `HttpException` afaik. I haven't dig deep into implementation code so I'm not sure if it might throw any other type of exception.